File: python/src/ttio/genomic_index.py
"""GenomicIndex — parallel per-read metadata for selective access.

Genomic analogue of :class:`ttio.acquisition_run.SpectrumIndex`. Loaded
eagerly when a :class:`ttio.genomic_run.GenomicRun` opens. Arrays have
length == read_count and are small enough to hold in memory; the heavy
signal channels (sequences, qualities) remain lazy on disk.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

import numpy as np

if TYPE_CHECKING:
    from .providers.base import StorageGroup


def _offsets_from_lengths(lengths: np.ndarray) -> np.ndarray:
    """Synthesize per-record byte offsets from a lengths array.

    ``offsets[i] = sum(lengths[0..i])``. Always uint64 to avoid the
    overflow cliff at >4 GB on deep WGS even when the input lengths
    array is uint32. Returns a fresh array of shape ``(n,)``; for
    empty input returns an empty uint64 array.

    v1.10 #10 (2026-05-04): canonical helper for the new on-disk
    schema where ``offsets`` is omitted from disk and computed at
    read time. Genomic + spectrum + chromatogram indexes share this
    helper.
    """
    n = int(lengths.shape[0])
    if n == 0:
        return np.zeros(0, dtype=np.uint64)
    out = np.empty(n, dtype=np.uint64)
    out[0] = 0
    if n > 1:
        # Use uint64 cumsum to avoid uint32 overflow on deep WGS.
        np.cumsum(lengths[:-1].astype(np.uint64, copy=False), out=out[1:])
    return out


def _read_offsets_from_lengths_dataset(
    idx_group: "StorageGroup",
    lengths_dtype: type = np.uint32,
) -> np.ndarray:
    """Read the ``lengths`` dataset and synthesize uint64 offsets.

    Convenience wrapper for callers that only need offsets — same as
    open + read + ``_offsets_from_lengths`` but avoids the boilerplate.
    """
    lengths = np.asarray(
        idx_group.open_dataset("lengths").read(), dtype=lengths_dtype,
    )
    return _offsets_from_lengths(lengths)


@dataclass(slots=True)
class GenomicIndex:
    """Parallel per-read arrays loaded eagerly at run open time.

    Genomic analogue of :class:`SpectrumIndex`. The arrays map 1:1 to
    the datasets under ``/study/genomic_runs/<name>/genomic_index/``
    (M82). They are small (length = read_count) and cheap to hold in
    memory; the heavy signal channels (sequences, qualities) remain
    lazy on disk.

    Parameters
    ----------
    offsets : numpy.ndarray
        ``uint64`` byte offset of each read into the ``sequences`` and
        ``qualities`` signal channels.
    lengths : numpy.ndarray
        ``uint32`` read length in bases for each read.
    chromosomes : list[str]
        Reference sequence name (e.g. ``"chr1"``) for each read.
        Variable-length strings stored as a compound VL_BYTES dataset
        on disk; held as a Python list in memory.
    positions : numpy.ndarray
        ``int64`` 0-based mapping position for each read.
    mapping_qualities : numpy.ndarray
        ``uint8`` mapping quality (Phred-scaled) for each read.
    flags : numpy.ndarray
        ``uint32`` SAM-style flags for each read. M82 uses UINT32 (not
        UINT16) to leave room for future extended flag bits beyond
        SAM's 12-bit range.

    Notes
    -----
    API status: Provisional (M82.1). Disk read/write methods are
    implemented in Task 6 of the M82.1 plan; until then they raise
    :class:`NotImplementedError`.

    Cross-language equivalents
    --------------------------
    Objective-C: ``TTIOGenomicIndex`` (M82.2 — to be implemented) ·
    Java: ``global.thalion.ttio.genomics.GenomicIndex`` (M82.3 — to
    be implemented).
    """

    offsets: np.ndarray            # uint64 — byte offset into sequence channel
    lengths: np.ndarray            # uint32 — read length in bases
    chromosomes: list[str]         # one per read
    positions: np.ndarray          # int64 — 0-based mapping position
    mapping_qualities: np.ndarray  # uint8
    flags: np.ndarray              # uint32

    @property
    def count(self) -> int:
        return int(self.offsets.shape[0])

    def indices_for_region(
        self, chromosome: str, start: int, end: int
    ) -> list[int]:
        """Return read indices on ``chromosome`` with start <= pos < end."""
        # TODO(M82): chromosome lookup is O(N) Python; vectorize once an
        # interned chromosome_ids column lands (BAM/CRAM-style id table).
        chrom_mask = np.array(
            [c == chromosome for c in self.chromosomes], dtype=bool
        )
        mask = chrom_mask & (self.positions >= start) & (self.positions < end)
        return np.where(mask)[0].tolist()

    def indices_for_unmapped(self) -> list[int]:
        """Return indices of unmapped reads (flag bit 0x4 set)."""
        return np.where(self.flags & 0x4)[0].tolist()

    def indices_for_flag(self, flag_mask: int) -> list[int]:
        """Return indices where ``(flags & flag_mask) != 0``."""
        return np.where(self.flags & flag_mask)[0].tolist()

    @classmethod
    def read(cls, idx_group: "StorageGroup") -> "GenomicIndex":
        """Load all columns from a ``genomic_index/`` StorageGroup.

        Chromosomes are stored as a uint16 id column + compound name
        lookup table (sibling datasets ``chromosome_ids`` and
        ``chromosome_names``). ``offsets`` is never on disk in v1.0+
        files — it's computed from ``cumsum(lengths)`` here.
        """
        from ttio import _hdf5_io as io

        lengths_ds = idx_group.open_dataset("lengths")
        positions_ds = idx_group.open_dataset("positions")
        mq_ds = idx_group.open_dataset("mapping_qualities")
        flags_ds = idx_group.open_dataset("flags")

        lengths = np.asarray(lengths_ds.read(), dtype=np.uint32)
        offsets = _offsets_from_lengths(lengths)
        positions = np.asarray(positions_ds.read(), dtype=np.int64)
        mapping_qualities = np.asarray(mq_ds.read(), dtype=np.uint8)
        flags = np.asarray(flags_ds.read(), dtype=np.uint32)

        ids_ds = idx_group.open_dataset("chromosome_ids")
        ids = np.asarray(ids_ds.read(), dtype=np.uint16)
        name_rows = io.read_compound_dataset(idx_group, "chromosome_names")
        name_table: list[str] = []
        for row in name_rows:
            v = row["name"]
            name_table.append(v.decode("utf-8") if isinstance(v, bytes) else v)
        chromosomes = [name_table[i] for i in ids.tolist()]

        return cls(
            offsets=offsets,
            lengths=lengths,
            chromosomes=chromosomes,
            positions=positions,
            mapping_qualities=mapping_qualities,
            flags=flags,
        )

    def write(self, idx_group: "StorageGroup") -> None:
        """Write all columns into ``idx_group``.

        v1.0 layout: ``chromosome_ids`` (uint16) + ``chromosome_names``
        (compound, one row per unique chromosome). Encounter-order id
        assignment — first occurrence of a name gets the next unused
        id. Cross-language byte-exact contract.

        ``offsets`` is never written — readers derive from
        ``cumsum(lengths)``.
        """
        from ttio import _hdf5_io as io
        from .enums import Precision

        io._write_uint32_channel(idx_group, "lengths", self.lengths, "gzip")
        io._write_int64_channel(idx_group, "positions", self.positions, "gzip")
        io._write_uint8_channel(
            idx_group, "mapping_qualities", self.mapping_qualities, "gzip"
        )
        io._write_uint32_channel(idx_group, "flags", self.flags, "gzip")

        name_to_id: dict[str, int] = {}
        names_in_order: list[str] = []
        ids = np.empty(len(self.chromosomes), dtype=np.uint16)
        for i, name in enumerate(self.chromosomes):
            slot = name_to_id.get(name)
            if slot is None:
                if len(names_in_order) > 65535:
                    raise ValueError(
                        "genomic_index: > 65,535 unique chromosome names; "
                        "uint16 chromosome_ids would overflow."
                    )
                slot = len(names_in_order)
                name_to_id[name] = slot
                names_in_order.append(name)
            ids[i] = slot

        ds = idx_group.create_dataset(
            "chromosome_ids", Precision.UINT16,
            length=int(ids.shape[0]),
            chunk_size=io.DEFAULT_SIGNAL_CHUNK,
            compression=io._compression_for("gzip"),
            compression_level=6,
        )
        ds.write(ids)
        io.write_compound_dataset(
            idx_group,
            "chromosome_names",
            [{"name": n} for n in names_in_order],
            [("name", io.vl_str())],
        )

File: java/src/main/java/global/thalion/ttio/genomics/GenomicIndex.java
/*
 * TTI-O Java Implementation
 * Copyright (C) 2026 DTW-Thalion
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */
package global.thalion.ttio.genomics;

import global.thalion.ttio.Enums.Compression;
import global.thalion.ttio.Enums.Precision;
import global.thalion.ttio.providers.CompoundField;
import global.thalion.ttio.providers.StorageDataset;
import global.thalion.ttio.providers.StorageGroup;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * Per-read offsets, lengths, positions, mapping qualities, flags, and
 * chromosome strings for one {@link GenomicRun}. Held in memory as
 * parallel arrays; loaded eagerly when the run is opened.
 *
 * <p>Genomic analogue of {@link global.thalion.ttio.SpectrumIndex}.</p>
 *
 * <p><b>API status:</b> Stable (v0.11 M82.3).</p>
 *
 * <p><b>Cross-language equivalents:</b> Objective-C
 * {@code TTIOGenomicIndex}, Python {@code ttio.genomic_index.GenomicIndex}.</p>
 */
public final class GenomicIndex {

    private static final int CHUNK_SIZE = 65536;
    private static final int COMPRESSION_LEVEL = 6;

    private final long[]   offsets;          // uint64 — byte offset into sequence channel
    private final int[]    lengths;          // uint32 — read length in bases
    private final List<String> chromosomes;  // one per read
    private final long[]   positions;        // int64 — 0-based mapping position
    private final byte[]   mappingQualities; // uint8
    private final int[]    flags;            // uint32

    public GenomicIndex(long[] offsets, int[] lengths,
                         List<String> chromosomes, long[] positions,
                         byte[] mappingQualities, int[] flags) {
        Objects.requireNonNull(offsets);
        Objects.requireNonNull(lengths);
        Objects.requireNonNull(chromosomes);
        Objects.requireNonNull(positions);
        Objects.requireNonNull(mappingQualities);
        Objects.requireNonNull(flags);
        if (lengths.length != offsets.length
                || chromosomes.size() != offsets.length
                || positions.length != offsets.length
                || mappingQualities.length != offsets.length
                || flags.length != offsets.length) {
            throw new IllegalArgumentException(
                "GenomicIndex column lengths must match");
        }
        this.offsets = offsets;
        this.lengths = lengths;
        this.chromosomes = List.copyOf(chromosomes);
        this.positions = positions;
        this.mappingQualities = mappingQualities;
        this.flags = flags;
    }

    /** Number of reads. */
    public int count() { return offsets.length; }

    /** Byte offset of read {@code i} in the sequences/qualities channels. */
    public long offsetAt(int i) { return offsets[i]; }
    /** Length of read {@code i} in bases. */
    public int lengthAt(int i) { return lengths[i]; }
    /** 0-based mapping position of read {@code i}. */
    public long positionAt(int i) { return positions[i]; }
    /** Phred-scaled mapping quality of read {@code i}, unsigned (0–255). */
    public int mappingQualityAt(int i) { return mappingQualities[i] & 0xFF; }
    /** SAM flags of read {@code i}. */
    public int flagsAt(int i) { return flags[i]; }
    /** Reference sequence name of read {@code i}. */
    public String chromosomeAt(int i) { return chromosomes.get(i); }

    /** Read indices on {@code chromosome} with {@code start <= position < end}. */
    public List<Integer> indicesForRegion(String chromosome, long start, long end) {
        List<Integer> out = new ArrayList<>();
        for (int i = 0; i < count(); i++) {
            if (chromosomes.get(i).equals(chromosome)
                    && positions[i] >= start && positions[i] < end) {
                out.add(i);
            }
        }
        return out;
    }

    /** Read indices where {@code (flags & 0x4) != 0}. */
    public List<Integer> indicesForUnmapped() { return indicesForFlag(0x4); }

    /** Read indices where {@code (flags & flagMask) != 0}. */
    public List<Integer> indicesForFlag(int flagMask) {
        List<Integer> out = new ArrayList<>();
        for (int i = 0; i < count(); i++) {
            if ((flags[i] & flagMask) != 0) out.add(i);
        }
        return out;
    }

    // ── Disk I/O via the StorageGroup protocol ─────────────────────

    /** Write this index into {@code idxGroup} (typically created via
     *  {@code parent.createGroup("genomic_index")}). The mathematically
     *  redundant {@code offsets} column is omitted; readers synthesize
     *  it from {@code cumsum(lengths)}.
     */
    public void writeTo(StorageGroup idxGroup) {
        writeInts (idxGroup, "lengths",          Precision.UINT32, lengths);
        writeLongs(idxGroup, "positions",        Precision.INT64,  positions);
        writeBytes(idxGroup, "mapping_qualities", Precision.UINT8, mappingQualities);
        writeInts (idxGroup, "flags",            Precision.UINT32, flags);

        // L1 (Task #82 Phase B.1, 2026-05-01): chromosomes are stored
        // as `chromosome_ids` (uint16) + `chromosome_names` (compound)
        // instead of a single VL-string compound. The old layout cost
        // 42 MB of HDF5 fractal-heap overhead per chr22 .tio file
        // (one heap block per chunk × 432 chunks) just to repeat one
        // byte-string 1.77M times. Encounter-order id assignment —
        // first occurrence gets the next unused id; cross-language
        // byte-exact contract.
        java.util.LinkedHashMap<String, Integer> nameToId = new java.util.LinkedHashMap<>();
        short[] ids = new short[chromosomes.size()];
        for (int i = 0; i < chromosomes.size(); i++) {
            String name = chromosomes.get(i);
            Integer slot = nameToId.get(name);
            if (slot == null) {
                if (nameToId.size() > 65535) {
                    throw new IllegalStateException(
                        "genomic_index: > 65,535 unique chromosome names; "
                        + "uint16 chromosome_ids would overflow.");
                }
                slot = nameToId.size();
                nameToId.put(name, slot);
            }
            ids[i] = slot.shortValue();
        }
        // Write chromosome_ids as uint16 (Java has no unsigned short,
        // so we pass the ids array via the StorageDataset writeAll
        // path which interprets the raw 16-bit pattern).
        StorageDataset cids;
        try {
            cids = idxGroup.createDataset(
                "chromosome_ids", Precision.UINT16, ids.length,
                CHUNK_SIZE, Compression.ZLIB, COMPRESSION_LEVEL);
        } catch (UnsupportedOperationException e) {
            cids = idxGroup.createDataset(
                "chromosome_ids", Precision.UINT16, ids.length,
                0, Compression.NONE, 0);
        }
        try (StorageDataset closeMe = cids) {
            closeMe.writeAll(ids);
        }
        // Write chromosome_names as compound[(name, VL_str)].
        List<CompoundField> nameFields = List.of(
            new CompoundField("name", CompoundField.Kind.VL_STRING));
        List<Object[]> nameRows = new ArrayList<>(nameToId.size());
        for (String n : nameToId.keySet()) nameRows.add(new Object[]{ n });
        try (StorageDataset ds = idxGroup.createCompoundDataset(
                "chromosome_names", nameFields, nameRows.size())) {
            ds.writeAll(nameRows);
        }
    }

    /** Read a {@link GenomicIndex} from an existing {@code genomic_index/}
     *  group (typically opened via {@code parent.openGroup("genomic_index")}).
     *
     *  <p>v1.10 #10 (2026-05-04): {@code offsets} is no longer stored
     *  on disk by default — it's mathematically derivable from
     *  {@code cumsum(lengths)}. Reader handles both layouts: pre-v1.10
     *  files have the column on disk (read directly); v1.10+ files
     *  synthesize it from lengths. Always uint64.</p>
     */
    @SuppressWarnings("unchecked")
    public static GenomicIndex readFrom(StorageGroup idxGroup) {
        int[]   lengths   = readInts (idxGroup, "lengths");
        long[]  offsets   = idxGroup.hasChild("offsets")
            ? readLongs(idxGroup, "offsets")
            : offsetsFromLengths(lengths);
        long[]  positions = readLongs(idxGroup, "positions");
        byte[]  mapqs     = readBytes(idxGroup, "mapping_qualities");
        int[]   flags     = readInts (idxGroup, "flags");

        // L1: read chromosome_ids (uint16) + chromosome_names
        // (compound) and materialise back to a List<String> so the
        // GenomicIndex API surface stays unchanged for callers.
        short[] ids;
        try (StorageDataset ds = idxGroup.openDataset("chromosome_ids")) {
            ids = (short[]) ds.readAll();
        }
        List<Object[]> nameRows;
        try (StorageDataset ds = idxGroup.openDataset("chromosome_names")) {
            nameRows = (List<Object[]>) ds.readAll();
        }
        List<String> nameTable = new ArrayList<>(nameRows.size());
        for (Object[] row : nameRows) {
            Object v = row[0];
            if (v instanceof byte[] b) {
                nameTable.add(new String(b, java.nio.charset.StandardCharsets.UTF_8));
            } else {
                nameTable.add(v == null ? "" : v.toString());
            }
        }
        List<String> chroms = new ArrayList<>(ids.length);
        for (short id : ids) {
            int idx = Short.toUnsignedInt(id);
            chroms.add(idx < nameTable.size() ? nameTable.get(idx) : "");
        }

        return new GenomicIndex(offsets, lengths, chroms, positions, mapqs, flags);
    }

    // ── Typed-channel helpers (chunked + zlib, falling back to raw) ─

    private static void writeLongs(StorageGroup g, String name, Precision p, long[] data) {
        StorageDataset ds;
        try {
            ds = g.createDataset(name, p, data.length, CHUNK_SIZE,
                    Compression.ZLIB, COMPRESSION_LEVEL);
        } catch (UnsupportedOperationException e) {
            ds = g.createDataset(name, p, data.length, 0, Compression.NONE, 0);
        }
        try (StorageDataset closeMe = ds) { closeMe.writeAll(data); }
    }

    private static void writeInts(StorageGroup g, String name, Precision p, int[] data) {
        StorageDataset ds;
        try {
            ds = g.createDataset(name, p, data.length, CHUNK_SIZE,
                    Compression.ZLIB, COMPRESSION_LEVEL);
        } catch (UnsupportedOperationException e) {
            ds = g.createDataset(name, p, data.length, 0, Compression.NONE, 0);
        }
        try (StorageDataset closeMe = ds) { closeMe.writeAll(data); }
    }

    private static void writeBytes(StorageGroup g, String name, Precision p, byte[] data) {
        StorageDataset ds;
        try {
            ds = g.createDataset(name, p, data.length, CHUNK_SIZE,
                    Compression.ZLIB, COMPRESSION_LEVEL);
        } catch (UnsupportedOperationException e) {
            ds = g.createDataset(name, p, data.length, 0, Compression.NONE, 0);
        }
        try (StorageDataset closeMe = ds) { closeMe.writeAll(data); }
    }

    private static long[] readLongs(StorageGroup g, String name) {
        try (StorageDataset ds = g.openDataset(name)) {
            return (long[]) ds.readAll();
        }
    }

    private static int[] readInts(StorageGroup g, String name) {
        try (StorageDataset ds = g.openDataset(name)) {
            return (int[]) ds.readAll();
        }
    }

    private static byte[] readBytes(StorageGroup g, String name) {
        try (StorageDataset ds = g.openDataset(name)) {
            return (byte[]) ds.readAll();
        }
    }

    /**
     * v1.10 #10 helper: synthesize per-record byte offsets from a
     * lengths array. {@code offsets[i] = sum(lengths[0..i])}, returned
     * as a {@code long[]} (i.e. uint64) to avoid the >4 GB overflow
     * cliff on deep WGS even though the input lengths are uint32.
     *
     * <p>Empty input returns {@code new long[0]}.</p>
     */
    public static long[] offsetsFromLengths(int[] lengths) {
        long[] out = new long[lengths.length];
        if (lengths.length == 0) return out;
        out[0] = 0L;
        long acc = 0L;
        for (int i = 1; i < lengths.length; i++) {
            // Mask up the int to its uint32 value before adding to long.
            acc += ((long) lengths[i - 1]) & 0xFFFFFFFFL;
            out[i] = acc;
        }
        return out;
    }
}
